src/comm.ts

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+"use strict";
+
+import { DisposableDelegate } from 'phosphor-disposable';
+
+import { 
+  IKernel, IKernelFuture, IKernelMessage, IKernelMessageOptions 
+} from './ikernel';
+
+import { createKernelMessage } from './kernel';
+
+import * as utils from './utils';
+
+
+/**
+ * A client side Comm interface.
+ */
+export
+interface IComm {
+  /**
+   * The uuid for the comm channel.
+   *
+   * Read-only
+   */
+  commId: string;
+
+  /** 
+   * The target name for the comm channel.
+   *
+   * Read-only
+   */
+  targetName: string;
+
+  /**
+   * The onClose handler.
+   */
+  onClose: (data?: any) => void;
+
+  /**
+   * The onMsg handler.
+   */
+  onMsg: (data: any) => void;
+
+  /**
+   * Open a comm with optional data.
+   */
+  open(data?: any, metadata?: any): IKernelFuture;
+
+  /**
+   * Send a comm message to the kernel.
+   */
+  send(data: any, metadata?: any, buffers?:(ArrayBuffer | ArrayBufferView)[]): IKernelFuture;
+
+  /**
+   * Close the comm.
+   */
+  close(data?: any, metadata?: any): IKernelFuture;
+}
+
+
+/**
+ * Contents of `comm_info` message.
+ */
+export
+interface ICommInfo {
+  /**
+   * Mapping of comm ids to target names.
+   */
+  comms: Map<string, string>;
+}
+
+
+/*
+ * CommManager for a Kernel.
+ *
+ * http://ipython.org/ipython-doc/dev/development/messaging.html#custom-messages
+ */
+export
+class CommManager {
+
+  /**
+   * Create a CommManager instance.
+   */
+  constructor(kernel: IKernel) {
+    this._kernel = kernel;
+    this._kernel.iopubReceived.connect(this._handleKernelMsg, this);
+  }
+
+  /**
+   * Connect to a comm, or create a new one.
+   *
+   * If a client-side comm already exists, it is returned.
+   */
+  connect(targetName: string, commId?: string): Promise<IComm> {
```

Comment:
A `Comm` created on the server side is asynchronous, because it my have to make an AMD load, so we treat everything on `CommManager` as asynchronous.